lib/internal/streams/pipeline.js

```diff
@@ -63,8 +69,84 @@ function popCallback(streams) {
   return streams.pop();
 }
 
+function isPromise(obj) {
+  return !!(obj && typeof obj.then === 'function');
+}
+
+function isReadable(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+function isWritable(obj) {
+  return !!(obj && typeof obj.write === 'function');
+}
+
+function isStream(obj) {
+  return isReadable(obj) || isWritable(obj);
+}
+
+function isIterable(obj, isAsync) {
+  if (!obj) return false;
+  if (isAsync === true) return !!obj[SymbolAsyncIterator];
+  if (isAsync === false) return !!obj[SymbolIterator];
+  return !!(obj[SymbolAsyncIterator] || obj[SymbolIterator]);
+}
+
+function makeAsyncIterable(val, name) {
+  if (isIterable(val)) {
+    return val;
+  } else if (isReadable(val)) {
+    return _fromReadable(val);
+  } else {
+    throw new ERR_INVALID_RETURN_VALUE(
+      'AsyncIterable, Readable', name, val);
+  }
+}
+
+async function* _fromPromise(val) {
+  yield await val;
+}
+
+async function* _fromReadable(val) {
+  if (isReadable(val)) {
+    // Legacy streams are not Iterable.
+
+    if (!Readable) {
+      Readable = require('_stream_readable');
+    }
+
+    try {
+      const it = Readable.prototype[SymbolAsyncIterator].call(val);
+      while (true) {
+        const { value, done } = await it.next();
+        if (done) return;
+        yield value;
+      }
```

Comment:
Am I missing something (which is totally possible) or is this loop effectively equivalent to `yield* it;`?

Reply:
Probably. I didn't even know you could do that. Learnt something new.

Reply:
Updated.